packages/financial-templates-lib/src/price-feed/TraderMadePriceFeed.js

```diff
@@ -0,0 +1,283 @@
+      return this.toBN(parseFixed(number.toString(), priceFeedDecimals).toString());
+    };
+  }
+
+  getCurrentPrice() {
+    return this.currentPrice;
+  }
+
+  getHistoricalPrice(time, verbose = false) {
+    if (this.lastUpdateTime === undefined) {
+      return null;
+    }
+
+    // Set first price time in `historicalPrices` to first non-null price.
+    let firstPriceTime;
+    for (let p in this.historicalPricesMinute) {
+      if (this.historicalPricesMinute[p] && this.historicalPricesMinute[p].openTime) {
+        firstPriceTime = this.historicalPricesMinute[p];
+        break;
+      }
+    }
+
+    // If there are no valid price time, return null.
+    if (!firstPriceTime) {
+      return null;
+    }
+
+    // If the time is before the first piece of data in the set, return null because
+    // the price is before the lookback window.
+    if (time < firstPriceTime.openTime) {
+      return null;
+    }
+
+    // historicalPrices are ordered from oldest to newest.
+    // This finds the first priceTime whose closeTime is after the provided time.
+    const match = this.historicalPricesMinute.find(price => {
+      return time < price.closeTime;
+    });
+
+    // If there is no match, that means that the time was past the last data point.
+    // In this case, the best match for this price is the current price.
+    if (match === undefined) {
+      let returnPrice = this.currentPrice;
+      if (verbose) {
+        console.group(`\n(${this.pair}) No OHLC available @ ${time}`);
+        console.log(
+          `- ✅ Time is later than earliest historical time, fetching current price: ${this.web3.utils.fromWei(
+            returnPrice.toString()
+          )}`
+        );
+        console.log(
+          `- ⚠️  If you want to manually verify the specific exchange prices, you can make a GET request to: \n- https://marketdata.tradermade.com/api/v1/live?currency=${this.pair}&api_key={api-key}`
+        );
+        console.groupEnd();
+      }
+      return this.currentPrice;
+    }
+
+    let returnPrice = match.closePrice;
+    if (verbose) {
+      console.group(`\n(${this.pair}) Historical OHLC @ ${match.closeTime}`);
+      console.log(`- ✅ Close Price:${this.web3.utils.fromWei(returnPrice.toString())}`);
+      console.log(
+        `- ⚠️  If you want to manually verify the specific exchange prices, you can make a GET request to: \n- https://marketdata.tradermade.com/api/v1/timeseries?currency=${this.pair}&api_key={api-key}&start_date=${time}&end_date=${match.closeTime}&format=records&interval=minute&period=${this.ohlcPeriod}`
+      );
+      console.groupEnd();
+    }
+    return returnPrice;
+  }
+
+  getHistoricalPricePeriods() {
+    return this.historicalPricesHourly;
+  }
+
+  getLastUpdateTime() {
+    return this.lastUpdateTime;
+  }
+
+  getLookback() {
+    return this.lookback;
+  }
+
+  getPriceFeedDecimals() {
+    return this.priceFeedDecimals;
+  }
+
+  async update() {
+    const currentTime = this.getTime();
```

Comment:
I'm going to suggest something that might be a pain in the butt here, but because some of these queries may be optional, it might be good to have a way to disable certain parts of the update.  Id suggest maybe breaking each one of these out into their own functions:
- updateLatest() -> sets current price and last update time
- updateMinute() -> sets minute history based on bot lookback
- updateHourly() -> sets hourly history based on affiliate lookback

Then in `update()` basically check if the params exist before calling them
```js
  await this.updateLatest()
  if(this.botLookback){
     await this.updateMinute()
  }
  if(this.affiliateLookback){
     await this.updateHourly()
  }
```

Reply:
updated